File: old/0101_SymmetricTree.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    def isSymmetric(self, root):
        if root == None:
            return True
        return self.sym(root.left, root.right)

    def sym(self, left, right):
        if left == None or right == None:
            return left == right

        if left.val != right.val:
            return False

        return self.sym(left.left, right.right) and self.sym(left.right, right.left)

File: top150/binary-tree-general/0101_Symmetric_Tree.go
// Leetcode: https://leetcode.com/problems/symmetric-tree/

package binarytreegeneral

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func isSymmetric(root *TreeNode) bool {
	return isSameTree(root, root)
}

func isSameTree(p *TreeNode, q *TreeNode) bool {
	if p == nil && q == nil {
		return true
	} else if (p != nil && q == nil) || (q != nil && p == nil) {
		return false
	}

	return (p.Val == q.Val) && isSameTree(p.Left, q.Right) && isSameTree(p.Right, q.Left)
}
